### PR TITLE
[ONNX] Add onnx::LayerNorm support for version 17

### PR DESCRIPTION
**Pull request by titaiwangms**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84293

### CODE REVIEW

**Comment by justinchuby** [comment]
I would add a test

**Comment by AllenTiTaiWang** [comment]
> I would add a test

We already have it .
https://github.com/pytorch/pytorch/blob/641c3952516baf444f75058f76cde59d0e1110f0/test/onnx/test_pytorch_onnx_onnxruntime.py#L3676-L3679

And this PR pass it.

**Review by justinchuby — APPROVED** [review]

File: torch/onnx/symbolic_opset17.py [hunk 1]
```diff
@@ -15,5 +15,19 @@
...
 """
 
+from torch.onnx import symbolic_helper
+
 # EDITING THIS FILE? READ THIS FIRST!
 # see Note [Edit Symbolic Files] in symbolic_helper.py
+
+
+@symbolic_helper.parse_args("v", "is", "v", "v", "f", "i")
+def layer_norm(g, input, normalized_shape, weight, bias, eps, cudnn_enable):
```
Comment (justinchuby) [inline torch/onnx/symbolic_opset17.py:25]: one more note: Please annotate the type and decorate the function with _beartype.beartype

Comment (justinchuby) [inline torch/onnx/symbolic_opset17.py:25]:
```suggestion
def layer_norm(g, input, normalized_shape, weight, bias, eps, cudnn_enable):
```

cudnn_enable is a boolean. Since it is not used, we can annotate as "none" in parse args

File: torch/onnx/symbolic_opset17.py [hunk 2]
```diff
@@ -15,5 +15,19 @@
...
+    return g.op(
+        "LayerNormalization",
+        input,
+        weight,
+        bias,
+        epsilon_f=eps,
+        axis_i=len(normalized_shape),
```
Comment (justinchuby) [inline torch/onnx/symbolic_opset17.py:32]: Can you write a brief comment inline explaining what axis is and why len(normalized_shape) is the same thing?

Comment (BowenBao) [reply torch/onnx/symbolic_opset17.py:32]: Good question, I think particularly in this test case, the value happens to match...

File: torch/onnx/symbolic_opset17.py (hunk 1 above)
Comment (AllenTiTaiWang) [reply torch/onnx/symbolic_opset17.py:25]: Should we use None or keep it as original (use v) in this case?

Comment (AllenTiTaiWang) [reply torch/onnx/symbolic_opset17.py:25]: If "none" is a common usage, I will add it into doc of parse args.

File: torch/onnx/symbolic_opset17.py (hunk 2 above)
Comment (AllenTiTaiWang) [reply torch/onnx/symbolic_opset17.py:32]: I was referencing on this:
https://github.com/pytorch/pytorch/blob/14093b5979cb5c0b777e3920819ab8252eb6d3ea/torch/onnx/symbolic_opset9.py#L2309

And according to https://github.com/onnx/onnx/blob/main/docs/Changelog.md#LayerNormalization-17
>axis : int (default is -1)
The first normalization dimension. If rank(X) is r, axis' allowed range is [-r, r]. Negative value means counting dimensions from the back.

I tried different params for layer_norm and input, but it works so far. @BowenBao is there a test case you have on mind that might fail?

Comment (BowenBao) [reply torch/onnx/symbolic_opset17.py:32]: What are the combinations you tried?
I suspect it should be `axis_i = -len(normalized_shape)`. However for this current test case, both `axis_i = 2` and `axis_i = -2` maps to the same dimension. So I'd recommend trying alternate the rank
```python
     model = torch.nn.LayerNorm([10, 10, 10]) 
     x = torch.randn(20, 5, 10, 10, 10) 
     self.run_test(model, x) 
```

Comment (AllenTiTaiWang) [reply torch/onnx/symbolic_opset17.py:32]: I see

File: torch/onnx/symbolic_opset17.py (hunk 1 above)
Comment (AllenTiTaiWang) [reply torch/onnx/symbolic_opset17.py]: offline meeting: more _beartype.beartype use case will be introduced in another PR from Justin. I have done annotation and beartype on shared layer norm function.

Comment (AllenTiTaiWang) [reply torch/onnx/symbolic_opset17.py]: Done

File: torch/onnx/symbolic_opset17.py (hunk 2 above)
Comment (AllenTiTaiWang) [reply torch/onnx/symbolic_opset17.py]: Done

**Review by BowenBao — APPROVED** [review]

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 3]
```diff
@@ -3674,9 +3674,12 @@ def forward(self, x):
         self.run_test(Model(), x)
 
     def test_layer_norm(self):
-        model = torch.nn.LayerNorm([10, 10])
-        x = torch.randn(20, 5, 10, 10)
-        self.run_test(model, x)
+        # As layer_norm works on the last D dimension, please keep
+        # this test case at least three dimension to prevent the
+        # situation of axis=2 mapping to the same axis as axis=-2
+        model2 = torch.nn.LayerNorm([10, 10, 10])
```
Comment (justinchuby) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:3680]: Any reason this is named model2? For clarity we can just do `layer_norm_model = ...` or just `model`

Comment (AllenTiTaiWang) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:3680]: Ahh, I was going to put two tests, and then only left one. I will clean it, and figure out the CI failure with beartype.

File: torch/onnx/symbolic_opset9.py [hunk 4]
```diff
@@ -2286,6 +2287,7 @@ def batch_norm(
         return res
 
 
+@_beartype.beartype
```
Comment (justinchuby) [inline torch/onnx/symbolic_opset9.py:2290]: You may remove the beartype check for this one here. I can handle it with #84091

File: torch/onnx/symbolic_opset17.py [hunk 5]
```diff
@@ -15,5 +15,35 @@
...
+    input: _C.Value,
+    normalized_shape: Sequence[int],
+    weight: _C.Value,
+    bias: _C.Value,
+    eps: float,
+    cudnn_enable: bool,
+):
+    # normalized_shape: input shape from an expected input of size
+    # axis: The first normalization dimension.
+    # layer_norm normalizes on the last D dimension,
```
Comment (justinchuby) [inline torch/onnx/symbolic_opset17.py:39]:
```suggestion
    # layer_norm normalizes on the last D dimensions,
```

File: torch/onnx/symbolic_opset17.py [hunk 6]
```diff
@@ -15,5 +15,35 @@
...
+    # which D is the size of normalized_shape
```
[the remark this answers is not in the document]
Comment (AllenTiTaiWang) [reply torch/onnx/symbolic_opset17.py]: Done

File: torch/onnx/symbolic_opset17.py (hunk 5 above)
Comment (AllenTiTaiWang) [reply torch/onnx/symbolic_opset17.py]: Done
